### PR TITLE
AGW: Envoy: plugin envoy proxy

### PR DESCRIPTION
**Pull request by pshelar**
## Summary

This patch adds proxy controller to process L7 traffic from
UE. This change is targeted for header enrichment feature.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

## Test Plan
Validate E2E HTTP flow on magma dev VM using manually configured envoy.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/service_manager.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/service_manager.py:332]: If we define it here I don't think we need this part?
```
            PROXY_TABLE: Tables(main_table=self.PROXY_TABLE_NUM,
                                type=ControllerType.SPECIAL),
```

File: lte/gateway/python/magma/pipelined/app/policy_mixin.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/policy_mixin.py]: check that test_enforcement_stats.py uses activate_rules directly so change the usage there as well

File: lte/gateway/python/magma/pipelined/app/policy_mixin.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/policy_mixin.py:358]: maybe

if self.proxy_controller and urls:

Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/policy_mixin.py:358]: The url check is done in he controller.

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py]: why is this commented out?

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py]: uncomment?

File: lte/gateway/python/magma/pipelined/app/enforcement_stats.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/enforcement_stats.py]: MSISDN

File: lte/gateway/python/magma/pipelined/openflow/registers.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/openflow/registers.py]: I think something like `set_in_port` is better as `reset_in_port` reads more like reset it to 0 to me

File: lte/gateway/python/magma/pipelined/policy_converters.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/policy_converters.py]: capital func name looks weird, typehint ipaddr and chane name to
ip_proto_to_ip_str maybe?

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py:40]: can we link to deign doc here orexpand this a bit

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py]: [nit] Header Enrichment app config:

File: lte/gateway/python/magma/pipelined/app/inout.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/inout.py]: [nit] Header Enrichment port

File: lte/gateway/python/magma/pipelined/tests/test_he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_he.py:190]: can you rewrite this test so that it uses the policyrules functions that call he_controller.get_subscriber_flows? 
Or just add another `test_ue_flows_enf(self)`, that way we can catch more issues than just calling controller directly

**Review by koolzz — CHANGES_REQUESTED** [review]
Also I think he.py would look better as header_enrichment, being a bit verbose here doesn't hurt.

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/he.py]: This will be done once envoy service is landed.

File: lte/gateway/python/magma/pipelined/tests/test_he.py
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/tests/test_he.py:190]: This test is just to validate the flows.
I will add separate test in enforcement test as another PR.

File: lte/gateway/python/magma/pipelined/app/inout.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/inout.py:109]: Should we just log something

File: lte/gateway/python/magma/pipelined/app/inout.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/inout.py]: I don't think this will work for CWF(you're using tunnel ids from activate flows req and maybe other things not in CWF) so lets just put it under ` if self.config.setup_type == 'LTE':`

File: lte/gateway/python/magma/pipelined/app/inout.py
Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/inout.py]: bump

File: lte/gateway/python/magma/pipelined/policy_converters.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/policy_converters.py:176]: either rename function to ipv4_address_to_str or return decoded_ip on both ip versions?

File: lte/gateway/python/magma/pipelined/tests/test_service_manager.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_service_manager.py:81]: Add a 
```
 self.assertEqual(
            self.service_manager.get_next_table_num(
                HeaderEnrichmentController.APP_NAME), 10)
```

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py:54]: Would this be better as a first LOGICAL table? I'll leave it up to you as proxy fits into both I think

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py:67]: Can you define public vars first, _ private vars last?also remove all the extra whitespace in between the vars

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/he.py:82]: maybe log?

**Review by koolzz — APPROVED** [review]
Approving but please address the style issues. Also I'm a bit confused how we're only passing the uplink tunnel to the he function, that can only set the tunnel into 1 direction, how about the other tunnel, is it somehow set by table 0?

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/he.py:54]: since it changes physical networking for packet flow, I am defining it as physical controller.

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/app/he.py:82]: init function prints config.

Comment (koolzz) [reply lte/gateway/python/magma/pipelined/app/he.py:82]: I meant because there is an DatapathLookupError

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/he.py]: `rule_no` => `rule_num`?

File: lte/gateway/python/magma/pipelined/app/enforcement.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/enforcement.py]: rather than passing None, why not define that parameter on the function as a default to `None`? this way you don't have to pass this to comply with the function signature

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/he.py]: It would be good to add a description to this `TODO` just to define what needs to be completed, looks more complete and documented

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/he.py]: do you really need this here?

File: lte/gateway/python/magma/pipelined/app/he.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/he.py]: this can be more descriptive specifying it as `match_in`, same for the one below as `match_out`

File: lte/gateway/python/magma/pipelined/app/policy_mixin.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/policy_mixin.py]: Initialized

File: lte/gateway/python/magma/pipelined/app/policy_mixin.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/policy_mixin.py:206]: missing uplink_tunnel param in the docs

File: lte/gateway/python/magma/pipelined/envoy_client.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/envoy_client.py]: `except grpc.RpcError`

File: lte/gateway/python/magma/pipelined/main.py
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/main.py:84]: why this overriding of the service configs? ideally the YAML files should be owning the configuration, not the other way around? can this be done using another variable?

**Review by ardzoht — CHANGES_REQUESTED** [review]

**Comment by ardzoht** [comment]
For styling and formatting, I'd recommend running a formatting tool like black or pyfmt to format the coding changes to be consistent or use the formatting including on the IDE.

File: lte/gateway/python/magma/pipelined/main.py
Comment (pshelar) [reply lte/gateway/python/magma/pipelined/main.py:84]: uplink_port is not part of config.   it is derived from ovs_uplink_port_name.
This is not new code, it is moved from inout controller to main, since two controller are going to use same config.
As you can see in the file plan is to move all config parsing in separate module from individual controllers. First step is to move to main.py file.

**Review by ardzoht — APPROVED** [review]

**Pull request merged**
